aten/src/ATen/native/SummaryOps.cpp

```diff
@@ -52,10 +51,77 @@ Tensor _bincount_cpu_template(
   }
   return output;
 }
+
+// A helper function to get the bin for histogram
+// while each bin is inclusive at the lower end and exclusive at the higher,
+// i.e. [start, end) the last bin is inclusive at both, i.e. [start, end], in
+// order to include maxvalue if exists
+template <typename input_t>
+inline int64_t getbin(input_t x, input_t min, input_t max, int64_t bins) {
+  if (x == max)
+    return bins - 1;
+  return (int64_t)((x - min) * bins / (max - min));
+}
+
+///////////////// histogram /////////////////
+template <typename input_t, typename weights_t>
+Tensor _histogram_cpu_template_uniformbins(
+    const Tensor& self,
+    int64_t nbins,
+    const Tensor& weights,
+    input_t min,
+    input_t max,
+    bool density) {
+
+  if (min == max) {
+    min -= 1;
+    max += 1;
+  }
+  TORCH_CHECK(nbins > 0, "bins must be > 0");
+  // Check if range is finite, hack to keep MSVC from complaining
+  TORCH_CHECK(
+      !(std::isinf((double)min) || std::isinf((double)max) ||
+        std::isnan((double)min) || std::isnan((double)max)),
+      "range of [",
+      min,
+      ", ",
+      max,
+      "] is not finite");
+  TORCH_CHECK(min < max, "max must be larger than min");
+
+  bool has_weights = weights.defined();
+  TORCH_CHECK(
+      !has_weights || weights.size(0) == self.size(0),
+      "input and weights should have the same length");
+
+  Tensor output;
+  int64_t self_size = self.size(0);
+
+  const input_t* self_p = self.data_ptr<input_t>();
+  if (has_weights) {
+    output = native::zeros({nbins}, weights.options());
+    weights_t* output_p = output.data_ptr<weights_t>();
+    const weights_t* weights_p = weights.data_ptr<weights_t>();
+    for (int64_t i = 0; i < self_size; i++) {
```

Comment:
Enumerating the tensor linearly and in a single thread like this won't be fast. 

You could write this with a custom parallel kernel, instead. You might also be able to write it using existing PyTorch operators. A custom parallel kernel seems like the fastest thing? Take a look at ATen/Parallel.h.

Reply:
Unfortunately the histc operator is still in th, but here's it's parallel bit:

https://github.com/pytorch/pytorch/blob/cb6c3526c6f21ad59aa1ee9cbe0d141c5f98c589/aten/src/TH/generic/THTensorMoreMath.cpp#L664

Reply:
Another option, which may be simpler and better, would be to flatten and sort the input, then binsearch the edges and infer the # of elements in each bin from their indices. This would use existing PyTorch operators that work in parallel for the hard bit (sorting). The binsearch could also be parallelized (one thread per bin), but that could probably be left for later.

Reply:
Sorry, forgot about the weights. I don't think we have a good mechanism to sort the weights at the same time as the input, unfortunately. If we did we could just multiply by them after identifying the bin edges. Darn.

Reply:
The CPU implementation isn't single-threaded on my machine, otherwise, the measured performance would be way worse, but it does rely on the compiler's auto-vectorization. It shouldn't be difficult to implement it as a unary operation with tensor iterators.

Reply:
I think it'd be interesting to try a version that explicitly parallel using TensorIterator.

Reply:
I had an opportunity to talk to my colleague and we're OK letting this be single-threaded, actually. Other similar operators in PyTorch are single-threaded today, and a multi-threaded implementation is a little tricky. Would you just add a comment that this is single threaded and consistent with similar operators?

Reply:
Will just add a comment about that in that case.